datumaro/plugins/synthetic_data/utils.py

```diff
@@ -0,0 +1,167 @@
+
+    @staticmethod
+    def process_nans(data):
+        nan_index = np.nonzero(np.isnan(data))
+        extend = np.array(range(nan_index[0][0] - 100, nan_index[0][0]))
+        delete_row = np.append(extend, nan_index)
+        return delete_row
+
+    def rescale(self, image_x, image_y, pad_x, pad_y):
+        xs = np.array(self.xs)
+        ys = np.array(self.ys)
+        if np.any(np.isnan(xs)):
+            delete_row = self.process_nans(xs)
+            xs = np.delete(xs, delete_row, axis=0)
+            ys = np.delete(ys, delete_row, axis=0)
+
+        if np.any(np.isnan(ys)):
+            delete_row = self.process_nans(ys)
+            xs = np.delete(xs, delete_row, axis=0)
+            ys = np.delete(ys, delete_row, axis=0)
+
+        if np.min(xs) < 0.0:
+            xs -= np.min(xs)
+        if np.min(ys) < 0.0:
+            ys -= np.min(ys)
+        xmax, xmin = np.max(xs), np.min(xs)
+        ymax, ymin = np.max(ys), np.min(ys)
+        self.xs = np.uint16(xs / (xmax - xmin + 1e-5) * (image_x - 2 * pad_x) + pad_x)
+        self.ys = np.uint16(ys / (ymax - ymin + 1e-5) * (image_y - 2 * pad_y) + pad_y)
+
+    def draw(self, image_x, image_y, draw_point, pad_x=6, pad_y=6):
+        self.rescale(image_x, image_y, pad_x, pad_y)
+        image = np.zeros((image_x, image_y), dtype=np.uint8)
+        for x, y in zip(self.xs, self.ys):
+            if draw_point:
+                image[x, y] = 127
+            else:
+                mask = "{:09b}".format(self._rng.randint(1, 511))
+                patch = 127 * np.array(list(map(int, list(mask))), dtype=np.uint8).reshape(3, 3)
+                image[x + 1 : x + 4, y + 1 : y + 4] = patch
+
+        return image
+
+
+def download_colorization_model(path):
+    proto_name = "colorization_deploy_v2.prototxt"
+    model_name = "colorization_release_v2.caffemodel"
+    npy_name = "pts_in_hull.npy"
+
+    if not osp.exists(osp.join(path, proto_name)):
+        url = "https://raw.githubusercontent.com/richzhang/colorization/caffe/colorization/models/"
+        proto = requests.get(url + proto_name)
+        open(osp.join(path, proto_name), "wb").write(proto.content)
+    if not osp.exists(osp.join(path, model_name)):
+        url = "http://eecs.berkeley.edu/~rich.zhang/projects/2016_colorization/files/demo_v2/"
+        model = requests.get(url + model_name)
+        open(osp.join(path, model_name), "wb").write(model.content)
+    if not osp.exists(osp.join(path, npy_name)):
+        url = "https://github.com/richzhang/colorization/raw/caffe/colorization/resources/"
+        pts_in_hull = requests.get(url + npy_name)
+        open(osp.join(path, npy_name), "wb").write(pts_in_hull.content)
+
+
+def rgb2lab(frame):
+    y_coeffs = np.array([0.212671, 0.715160, 0.072169], dtype=np.float32)
+    frame = np.where(frame > 0.04045, np.power((frame + 0.055) / 1.055, 2.4), frame / 12.92)
+    y = frame @ y_coeffs.T
+    L = np.where(y > 0.008856, 116 * np.cbrt(y) - 16, 903.3 * y)
+    return L
```

Comment:
Is it better than just using OpenCV? `cv2.cvtColor(image, cv2.COLOR_BGR2LAB)`

Reply:
Initially we used this function, but in a multiprocessor program it sometimes freezes

Reply:
It looks suspicious, because with multiprocessing we have separate processes, which should not have memory conflicts. I think, it should be reproduced and reported to the OpenCV team.

Reply:
cv.COLOR_Lab2BGR works fine